### PR TITLE
A bunch of refactorings

### PR DESCRIPTION
**Pull request by jackfranklin**
Let me know if you have any Qs :)

### CODE REVIEW

File: Main.elm [hunk 1]
```diff
@@ -14,7 +14,11 @@ import List.Extra
 
 
 type alias Model =
-  List ( String, String )
+  List Param
+
+
+type alias Param =
+  ( String, String )
```
Comment (jackfranklin) [inline Main.elm:21]: Doing these changes means everything is a bit more self documenting and if the concept if a `Param` changes we only have to change one type defn.

File: Main.elm [hunk 2]
```diff
@@ -51,30 +55,50 @@ main =
...
 
+withBlankDefault : Maybe Param -> Param
+withBlankDefault =
+  Maybe.withDefault ( "", "" )
+
+
+safeGetAtIndex : Model -> Int -> Param
+safeGetAtIndex model index =
+  List.Extra.getAt model index
+    |> withBlankDefault
```
Comment (jackfranklin) [inline Main.elm:66]: Both of these I would consider in a bigger proj moving out of `Main.elm` and into modules that group up your extra functions.

File: Main.elm [hunk 3]
```diff
@@ -51,30 +55,50 @@ main =
...
   case action of
     UpdateKey ( index, newKey ) ->
       let
-        current =
-          List.Extra.getAt model index
-            |> Maybe.withDefault ( "", "" )
-
-        new =
-          ( newKey, snd current )
+        ( _, value ) =
```
Comment (jackfranklin) [inline Main.elm:74]: In Elm we can destructure in assignments just like in ES2105, but also use `_` to denote that we don't care about that particular value.

Comment (jackfranklin) [inline Main.elm:74]: I do this a lot with tuples to save `fst` and `snd` calls

File: Main.elm [hunk 4]
```diff
@@ -51,30 +55,50 @@ main =
...
-          ( fst current, newValue )
+        ( key, _ ) =
+          safeGetAtIndex model index
       in
-        updateIn model index new
+        updateIn model index ( key, newValue )
+
+
+createInput : Signal.Address Action -> ( Int, Param ) -> Html
+createInput address ( index, ( param, paramValue ) ) =
```
Comment (jackfranklin) [inline Main.elm:88]: destructuring in the args tidies up the body :)

File: Main.elm [hunk 5]
```diff
@@ -84,24 +108,8 @@ view address model =
...
-                      ]
-                      []
-                  , input
-                      [ value (snd tuple)
-                      , on "input" targetValue (\str -> Signal.message address (UpdateValue ( index, str )))
-                      ]
-                      []
-                  ]
-              )
+          |> List.map (createInput address)
```
Comment (jackfranklin) [inline Main.elm:112]: Now `createInput` is pulled out I can partially apply it with the `address` and use that to map over the results.

**Comment by jackfranklin** [comment]
Worth noting all of this is very minor - great job :)

**Comment by OliverJAsh** [comment]
Awesome! :+1: Thank you so much!

My next step is to figure out how to perform side effects using ports. This will eventually be used as part of a Chrome extension (to replace [my old one](https://github.com/OliverJAsh/url-chrome/blob/master/popup.js)) that updates the URL when you modify the query parameters in the table of inputs.

I want to see if I can use ports to console log the model every time it changes. To do that I think I need to update my `update` function to return `Effect`s, is that right?

**Pull request merged**